File: ml/scan/pipeline/lib/src/fuse_passes.dart
/// Fusion de deux passes OCR sur la même image, ligne physique à ligne.
/// Portage de référence de `research/reference/fuse_passes.py`.
///
/// Les deux passes (brute, prétraitée) abîment des lignes différentes : l'une
/// colle une lettre au prix, l'autre saute un article. Alignées par position
/// verticale, elles se complètent : une ligne non chiffrée dans la passe
/// principale prend la lecture de l'autre, une ligne absente est insérée, un
/// montant différent devient une alternative que le décodeur sous contrainte
/// arbitre. Rien n'est inventé : chaque montant vient d'une passe OCR.
library;

import 'lines.dart';
import 'structure.dart';

const double alignmentToleranceRatio = 0.6;

class FusedPass {
  const FusedPass({required this.lines, this.alternatives = const {}});

  final List<PhysicalLine> lines;

  /// Index de ligne fusionnée → montant alternatif en centimes.
  final Map<int, int> alternatives;
}

class _Placed {
  const _Placed(this.line, this.centerY, [this.alternativeCents]);

  final PhysicalLine line;
  final double centerY;
  final int? alternativeCents;
}

double _centerY(PhysicalLine line) {
  final centers = [for (final word in line.words) word.centerY]..sort();
  return centers[centers.length ~/ 2];
}

double _medianHeight(List<PhysicalLine> lines) {
  final heights = [
    for (final line in lines)
      for (final word in line.words) word.height,
  ]..sort();
  return heights.isEmpty ? 1.0 : heights[heights.length ~/ 2];
}

int? _cents(PhysicalLine line) {
  final priced = rightmostPrice(line);
  return priced == null ? null : (priced.price * 100).round();
}

int _priceCount(PhysicalLine line) =>
    line.words.where((word) => parsePrice(word.text) != null).length;

(Map<int, List<int>>, List<int>) _matchSecondary(
  List<PhysicalLine> primary,
  List<PhysicalLine> secondary,
  double tolerance,
) {
  final matches = {for (var i = 0; i < primary.length; i++) i: <int>[]};
  final unmatched = <int>[];
  final centers = [for (final line in primary) _centerY(line)];
  for (final (secondaryIndex, line) in secondary.indexed) {
    final center = _centerY(line);
    int? nearest;
    for (var i = 0; i < primary.length; i++) {
      if (nearest == null ||
          (centers[i] - center).abs() < (centers[nearest] - center).abs()) {
        nearest = i;
      }
    }
    if (nearest == null || (centers[nearest] - center).abs() > tolerance) {
      unmatched.add(secondaryIndex);
    } else {
      matches[nearest]!.add(secondaryIndex);
    }
  }
  return (matches, unmatched);
}

List<_Placed> _placeSingle(PhysicalLine primary, PhysicalLine match) {
  final primaryCents = _cents(primary);
  final matchCents = _cents(match);
  if (primaryCents == null && matchCents != null) {
    return [_Placed(match, _centerY(match))];
  }
  final alternative =
      primaryCents != null && matchCents != null && matchCents != primaryCents
      ? matchCents
      : null;
  return [_Placed(primary, _centerY(primary), alternative)];
}

List<_Placed> _placeSeveral(PhysicalLine primary, List<PhysicalLine> matches) {
  if (_priceCount(primary) >= 2) {
    return [for (final match in matches) _Placed(match, _centerY(match))];
  }
  final primaryCents = _cents(primary);
  return [
    _Placed(primary, _centerY(primary)),
    for (final match in matches)
      if (_cents(match) case final cents? when cents != primaryCents)
        _Placed(match, _centerY(match)),
  ];
}

FusedPass fusePasses(List<PhysicalLine> primary, List<PhysicalLine> secondary) {
  final tolerance = alignmentToleranceRatio * _medianHeight(primary);
  final (matches, unmatched) = _matchSecondary(primary, secondary, tolerance);
  final placed = <_Placed>[];
  for (final (index, line) in primary.indexed) {
    final matched = [for (final i in matches[index]!) secondary[i]];
    if (matched.isEmpty) {
      placed.add(_Placed(line, _centerY(line)));
    } else if (matched.length == 1) {
      placed.addAll(_placeSingle(line, matched.single));
    } else {
      placed.addAll(_placeSeveral(line, matched));
    }
  }
  placed.addAll([
    for (final i in unmatched) _Placed(secondary[i], _centerY(secondary[i])),
  ]);
  final ordered = [for (final (i, entry) in placed.indexed) (i, entry)]
    ..sort((a, b) {
      final byCenter = a.$2.centerY.compareTo(b.$2.centerY);
      return byCenter != 0 ? byCenter : a.$1.compareTo(b.$1);
    });
  return FusedPass(
    lines: [for (final entry in ordered) entry.$2.line],
    alternatives: {
      for (final (index, entry) in ordered.indexed)
        index: ?entry.$2.alternativeCents,
    },
  );
}

File: ml/scan/research/reference/fuse_passes.py
"""Fusion de deux passes OCR sur la même image, ligne physique à ligne.

Les deux passes (brute, prétraitée) abîment des lignes différentes : l'une
colle une lettre au prix, l'autre saute un article. Alignées par position
verticale, elles se complètent : une ligne non chiffrée dans la passe
principale prend la lecture de l'autre, une ligne absente est insérée, un
montant différent devient une alternative que le décodeur sous contrainte
arbitre. Rien n'est inventé : chaque montant vient d'une passe OCR.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from reference.lines import PhysicalLine
from reference.structure import _rightmost_price, parse_price

ALIGNMENT_TOLERANCE_RATIO = 0.6


@dataclass(frozen=True)
class FusedPass:
    lines: list[PhysicalLine]
    alternatives: dict[int, int] = field(default_factory=dict)


@dataclass(frozen=True)
class _Placed:
    line: PhysicalLine
    center_y: float
    alternative_cents: int | None = None


def _center_y(line: PhysicalLine) -> float:
    centers = sorted(word.center_y for word in line.words)
    return centers[len(centers) // 2]


def _median_height(lines: list[PhysicalLine]) -> float:
    heights = sorted(word.height for line in lines for word in line.words)
    return heights[len(heights) // 2] if heights else 1.0


def _cents(line: PhysicalLine) -> int | None:
    priced = _rightmost_price(line)
    return None if priced is None else round(priced[0] * 100)


def _price_count(line: PhysicalLine) -> int:
    return sum(parse_price(word.text) is not None for word in line.words)


def _match_secondary(
    primary: list[PhysicalLine], secondary: list[PhysicalLine], tolerance: float
) -> tuple[dict[int, list[int]], list[int]]:
    matches: dict[int, list[int]] = {index: [] for index in range(len(primary))}
    unmatched: list[int] = []
    centers = [_center_y(line) for line in primary]
    for secondary_index, line in enumerate(secondary):
        center = _center_y(line)
        nearest = min(
            range(len(primary)), key=lambda i: abs(centers[i] - center), default=None
        )
        if nearest is None or abs(centers[nearest] - center) > tolerance:
            unmatched.append(secondary_index)
        else:
            matches[nearest].append(secondary_index)
    return matches, unmatched


def _place_single(primary: PhysicalLine, match: PhysicalLine) -> list[_Placed]:
    primary_cents = _cents(primary)
    match_cents = _cents(match)
    if primary_cents is None and match_cents is not None:
        return [_Placed(match, _center_y(match))]
    alternative = (
        match_cents
        if primary_cents is not None
        and match_cents is not None
        and match_cents != primary_cents
        else None
    )
    return [_Placed(primary, _center_y(primary), alternative)]


def _place_several(primary: PhysicalLine, matches: list[PhysicalLine]) -> list[_Placed]:
    if _price_count(primary) >= 2:
        return [_Placed(match, _center_y(match)) for match in matches]
    primary_cents = _cents(primary)
    placed = [_Placed(primary, _center_y(primary))]
    placed.extend(
        _Placed(match, _center_y(match))
        for match in matches
        if _cents(match) not in (None, primary_cents)
    )
    return placed


def fuse_passes(
    primary: list[PhysicalLine], secondary: list[PhysicalLine]
) -> FusedPass:
    tolerance = ALIGNMENT_TOLERANCE_RATIO * _median_height(primary)
    matches, unmatched = _match_secondary(primary, secondary, tolerance)
    placed: list[_Placed] = []
    for index, line in enumerate(primary):
        matched = [secondary[i] for i in matches[index]]
        if not matched:
            placed.append(_Placed(line, _center_y(line)))
        elif len(matched) == 1:
            placed.extend(_place_single(line, matched[0]))
        else:
            placed.extend(_place_several(line, matched))
    placed.extend(_Placed(secondary[i], _center_y(secondary[i])) for i in unmatched)
    placed.sort(key=lambda entry: entry.center_y)
    return FusedPass(
        lines=[entry.line for entry in placed],
        alternatives={
            index: entry.alternative_cents
            for index, entry in enumerate(placed)
            if entry.alternative_cents is not None
        },
    )
